File: core/src/main/java/javabot/web/resources/AdminResource.java
package javabot.web.resources;

import com.google.inject.Injector;
import io.dropwizard.views.View;
import io.dropwizard.views.freemarker.FreemarkerViewRenderer;
import javabot.Javabot;
import javabot.dao.AdminDao;
import javabot.dao.ApiDao;
import javabot.dao.ChannelDao;
import javabot.dao.ConfigDao;
import javabot.model.Admin;
import javabot.model.ApiEvent;
import javabot.model.Channel;
import javabot.model.Config;
import javabot.model.EventType;
import javabot.web.auth.Restricted;
import javabot.web.model.Authority;
import javabot.web.model.User;
import javabot.web.views.AdminIndexView;
import javabot.web.views.ChannelEditView;
import javabot.web.views.ConfigurationView;
import javabot.web.views.JavadocView;
import org.bson.types.ObjectId;

import javax.inject.Inject;
import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.Consumes;
import javax.ws.rs.FormParam;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.Response.Status;
import java.net.URI;

@Path("/admin")
public class AdminResource {
    @Inject
    private Injector injector;

    @Inject
    private AdminDao adminDao;

    @Inject
    private ApiDao apiDao;

    @Inject
    private ConfigDao configDao;

    @Inject
    private ChannelDao channelDao;

    @Inject
    private Javabot javabot;

    @GET
    public View index(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user) {
        return new AdminIndexView(injector, request);
    }

    @GET
    @Path("/config")
    public View config(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user) {
        return new ConfigurationView(injector, request);
    }

    @GET
    @Path("/javadoc")
    public View javadoc(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user) {
        return new JavadocView(injector, request);
    }

    @GET
    @Path("/newChannel")
    public View newChannel(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user) {
        return new ChannelEditView(injector, request, new Channel());
    }

    @POST
    @Path("/saveConfig")
    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
    public View saveConfig(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user,
                           @FormParam("server") String server, @FormParam("url") String url, @FormParam("port") Integer port,
                           @FormParam("historyLength") Integer historyLength, @FormParam("trigger") String trigger,
                           @FormParam("nick") String nick, @FormParam("password") String password,
                           @FormParam("throttleThreshold") Integer throttleThreshold,
                           @FormParam("minimumNickServAge") Integer minimumNickServAge) {
        Config config = configDao.get();
        config.setServer(server);
        config.setUrl(url);
        config.setPort(port);
        config.setHistoryLength(historyLength);
        config.setTrigger(trigger);
        config.setNick(nick);
        config.setPassword(password);
        config.setThrottleThreshold(throttleThreshold);
        config.setMinimumNickServAge(minimumNickServAge);
        configDao.save(config);
        return new ConfigurationView(injector, request);
    }

    @GET
    @Path("/enableOperation/{name}")
    public View enableOperation(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user,
                                @PathParam("name") String name) {
        javabot.enableOperation(name);
        return new ConfigurationView(injector, request);
    }

    @GET
    @Path("/disableOperation/{name}")
    public View disableOperation(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user,
                                 @PathParam("name") String name) {
        javabot.disableOperation(name);
        return new ConfigurationView(injector, request);
    }


    @GET
    @Path("/delete/{id}")
    public View deleteAdmin(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user,
                            @PathParam("id") String id) {
        Admin admin = adminDao.find(new ObjectId(id));
        if (admin != null && !admin.getBotOwner()) {
            adminDao.delete(admin);
        }
        return index(request, user);
    }

    @POST
    @Path("/add")
    public View addAdmin(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user,
                         @FormParam("ircName") String ircName, @FormParam("hostName") String hostName,
                         @FormParam("emailAddress") String emailAddress) {
        Admin admin = adminDao.getAdminByEmailAddress(emailAddress);
        if (admin == null) {
            admin = new Admin();
            admin.setIrcName(ircName);
            admin.setHostName(hostName);
            admin.setEmailAddress(emailAddress);
            adminDao.save(admin);
        }
        return index(request, user);
    }

    @POST
    @Path("/addApi")
    public View addApi(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user,
                       @FormParam("name") String name, @FormParam("baseUrl") String baseUrl,
                       @FormParam("downloadUrl") String downloadUrl) {
        apiDao.save(new ApiEvent(user.getEmail(), name, baseUrl, downloadUrl));
        return index(request, user);
    }

    @GET
    @Path("/deleteApi/{id}")
    public View deleteApi(@Context HttpServletRequest request,
                          @Restricted(Authority.ROLE_ADMIN) User user,
                          @PathParam("id") String id) {
        apiDao.save(new ApiEvent(EventType.DELETE, user.getEmail(), new ObjectId(id)));
        return index(request, user);
    }

    @POST
    @Path("/saveChannel")
    public View saveChannel(@Context HttpServletRequest request, @Restricted(Authority.ROLE_ADMIN) User user,
                            @FormParam("id") String id, @FormParam("name") String name, @FormParam("key") String key,
                            @FormParam("logged") boolean logged) {
        Channel channel = id == null ? new Channel(name, key, logged) : new Channel(new ObjectId(id), name, key, logged);
        channelDao.save(channel);
        return index(request, user);
    }

}

File: src/main/kotlin/javabot/web/resources/AdminResource.kt
package javabot.web.resources

import io.quarkus.qute.TemplateInstance
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.servlet.http.HttpServletRequest
import jakarta.ws.rs.Consumes
import jakarta.ws.rs.FormParam
import jakarta.ws.rs.GET
import jakarta.ws.rs.POST
import jakarta.ws.rs.Path
import jakarta.ws.rs.PathParam
import jakarta.ws.rs.WebApplicationException
import jakarta.ws.rs.core.Context
import jakarta.ws.rs.core.MediaType
import javabot.Javabot
import javabot.JavabotConfig
import javabot.dao.AdminDao
import javabot.dao.ApiDao
import javabot.dao.ChannelDao
import javabot.dao.ConfigDao
import javabot.model.Admin
import javabot.model.ApiEvent
import javabot.model.Channel
import javabot.model.javadoc.JavadocApi
import javabot.web.auth.Restricted
import javabot.web.model.Authority
import javabot.web.model.User
import javabot.web.views.TemplateService
import org.bson.types.ObjectId

@Path("/admin")
@ApplicationScoped
class AdminResource
@Inject
constructor(
    var templateService: TemplateService,
    var adminDao: AdminDao,
    var apiDao: ApiDao,
    var configDao: ConfigDao,
    var channelDao: ChannelDao,
    var javabot: Javabot,
    var config: JavabotConfig,
) {

    @GET
    fun index(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
    ): TemplateInstance {
        val current = adminDao.getAdminByEmailAddress(user.email)
        return if (current == null) templateService.createError403View()
        else templateService.createAdminIndexView(request, current, Admin())
    }

    @GET
    @Path("/config")
    fun config(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
    ): TemplateInstance {
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        return templateService.createConfigurationView(request)
    }

    @GET
    @Path("/javadoc")
    fun javadoc(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
    ): TemplateInstance {
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        return templateService.createJavadocAdminView(request)
    }

    @GET
    @Path("/newChannel")
    fun newChannel(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
    ): TemplateInstance {
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        return templateService.createChannelEditView(request, Channel())
    }

    @GET
    @Path("/editChannel/{channel}")
    fun editChannel(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @PathParam("channel") channel: String,
    ): TemplateInstance {

        // TODO redirect to / if channel is null
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        return templateService.createChannelEditView(request, channelDao.get(channel)!!)
    }

    @POST
    @Path("/saveChannel")
    fun saveChannel(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @FormParam("id") id: String?,
        @FormParam("name") name: String,
        @FormParam("key") key: String,
        @FormParam("logged") logged: Boolean,
    ): TemplateInstance {
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        val channel =
            if (id == null) Channel(name, key, logged) else Channel(ObjectId(id), name, key, logged)
        channelDao.save(channel)
        return index(request, user)
    }

    @POST
    @Path("/saveConfig")
    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
    fun saveConfig(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @FormParam("server") server: String,
        @FormParam("url") url: String,
        @FormParam("port") port: Int,
        @FormParam("historyLength") historyLength: Int,
        @FormParam("trigger") trigger: String,
        @FormParam("nick") nick: String,
        @FormParam("password") password: String,
        @FormParam("throttleThreshold") throttleThreshold: Int,
        @FormParam("minimumNickServAge") minimumNickServAge: Int,
    ): TemplateInstance {
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        val config = configDao.get()
        config.server = server
        config.url = url
        config.port = port
        config.historyLength = historyLength
        config.trigger = trigger
        config.nick = nick
        config.password = password
        config.throttleThreshold = throttleThreshold
        config.minimumNickServAge = minimumNickServAge
        configDao.save(config)
        return templateService.createConfigurationView(request)
    }

    @GET
    @Path("/enableOperation/{name}")
    fun enableOperation(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @PathParam("name") name: String,
    ): TemplateInstance {
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        javabot.enableOperation(name)
        return templateService.createConfigurationView(request)
    }

    @GET
    @Path("/disableOperation/{name}")
    fun disableOperation(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @PathParam("name") name: String,
    ): TemplateInstance {
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        javabot.disableOperation(name)
        return templateService.createConfigurationView(request)
    }

    @GET
    @Path("/edit/{id}")
    fun editAdmin(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @PathParam("id") id: String,
    ): TemplateInstance {
        val current =
            adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)

        return templateService.createAdminIndexView(request, current, adminDao.find(ObjectId(id)))
    }

    @GET
    @Path("/delete/{id}")
    fun deleteAdmin(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @PathParam("id") id: String,
    ): TemplateInstance {
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        val admin = adminDao.find(ObjectId(id))
        if (admin != null && (!admin.botOwner)) {
            adminDao.delete(admin)
        }
        return index(request, user)
    }

    @POST
    @Path("/add")
    fun addAdmin(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @FormParam("ircName") ircName: String,
        @FormParam("hostName") hostName: String,
        @FormParam("emailAddress") emailAddress: String,
    ): TemplateInstance {
        var admin: Admin? = adminDao.getAdminByEmailAddress(emailAddress)
        if (admin == null) {
            admin = Admin(ircName, emailAddress, hostName, true)
        } else {
            admin.ircName = ircName
            admin.hostName = hostName
            admin.emailAddress = emailAddress
        }
        adminDao.save(admin)
        return index(request, user)
    }

    @POST
    @Path("/addApi")
    fun addApi(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @FormParam("name") name: String?,
        @FormParam("groupId") groupId: String?,
        @FormParam("artifactId") artifactId: String?,
        @FormParam("version") version: String?,
    ): TemplateInstance {

        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        version?.let {
            val apiName = name ?: artifactId ?: throw WebApplicationException(400)
            val api = JavadocApi(config, apiName, groupId ?: "", artifactId ?: "", version)
            apiDao.save(api)
            apiDao.save(ApiEvent.add(user.email, api))
        }

        return javadoc(request, user)
    }

    @GET
    @Path("/deleteApi/{id}")
    fun deleteApi(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @PathParam("id") id: String,
    ): TemplateInstance {
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        apiDao.delete(ObjectId(id))
        return javadoc(request, user)
    }

    @GET
    @Path("/reloadApi/{id}")
    fun reloadApi(
        @Context request: HttpServletRequest,
        @Restricted(Authority.ROLE_ADMIN) user: User,
        @PathParam("id") id: String,
    ): TemplateInstance {
        adminDao.getAdminByEmailAddress(user.email) ?: throw WebApplicationException(403)
        apiDao.find(ObjectId(id))?.let { apiDao.save(ApiEvent.reload(user.email, it)) }
        return javadoc(request, user)
    }
}
